src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.AnalyzerConfig;
+using AnalyzerOptions = System.Collections.Immutable.ImmutableDictionary<string, string>;
+using TreeOptions = System.Collections.Immutable.ImmutableDictionary<string, Microsoft.CodeAnalysis.ReportDiagnostic>;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a set of <see cref="AnalyzerConfig"/> that can be applied to a given source location.
+    /// </summary>
+    public sealed class AnalyzerConfigSet
```

Comment:
I still think this should be a `readonly struct`.

Reply:
My thought here is given this is a type that will get maybe one allocation _per project, for the lifetime of Visual Studio_, it's unclear to me why we're concerned about allocations, especially given the behavior then gets strange. If we want to add additional caching under the covers here, that might make us regret making this a struct because then we'll have multiple fields, and that that means for threadsafety, etc.

Reply:
Consider too the IDE will have an `AsyncLazy<AnalyzerConfigSet>`, and if you make it a struct it'll just have to JIT a custom copy rather than reusing the reference __Canon one, thereby increasing JIT time, start time, and probably offsetting the memory gained by this with the JITted method overhead. That type is _not_ small.